### ISSUE TITLE
Dynamic references using custom database or schema

### ISSUE THREAD

**Issue by mroy-seedbox** [dbt-labs/dbt-core#2921]
### Describe the bug
**tl;dr: It's impossible to use a model's database or schema to dynamically generate the name of a reference.**

Let's say I have two models: `test_db1` & `test_db2` (where db1 & db2 are the names of databases), and other models referencing those. I can easily reference these models directly, using `{{ ref('test_db1') }}`. However, if I want to dynamically generate the name of the reference using a model's database (or schema) (like this: `{{ ref('test_' + database) }}`), it may generate the wrong reference and produce an error:

**Situation 1:**
- If `database` is the same as the profile's target database (no custom database is used), everything will be okay.
- Ex: target database: `db1`, `model_x` on `db1`.
- This will work and reference `test_db1`.

**Situation 2:**
- If `database` is not the same as the profile's target database (a custom database is used), the wrong reference will be used.
- Ex: target database: `db1`, `model_x` on `db2`
- This will produce a compilation error, because during the parsing/manifest phase, the reference was to `test_db1`, but during the compilation/run phase, the reference switched to `test_db2`.
- Unfortunately, the proposed fix is not helpful in this case; it forces us to hardcode the reference and prevents us from generating it dynamically.

```
Encountered an error:
Runtime Error
  Compilation Error in model model_x (path/to/model_x.sql)
    dbt was unable to infer all dependencies for the model "model_x".
    This typically happens when ref() is placed within a conditional block.
    
    To fix this, add the following hint to the top of the model "model_x":
    
    -- depends_on: {{ ref('test_db2') }}
    
    > in macro ref (macros/ref.sql)
    > called by model model_x (path/to/model_x.sql)
```

**Situation 3:**
- If `database` is not the same as the profile's target database (a custom database is used), a nonexistent reference will be used.
- Ex: target database: `dbz`, `model_x` on `db2`
- This will produce a compilation error, because during the parsing/manifest phase, the reference was to `test_dbz`, which doesn't exist.
- Dbt will never reach the compilation/run phase, where the reference would have switched to `test_db2`.

```
Encountered an error:
Compilation Error in model model_x (path/to/model_x.sql)
  Model 'model.project.model_x' (path/to/model_x.sql) depends on a node named 'test_dbz' which was not found
```

The problem seems to be that during the parsing/manifest phase (as described in the [`execute` documentation](https://docs.getdbt.com/reference/dbt-jinja-functions/execute/)), a model's database and schema are not resolved yet - they always refer to the profile's target database & schema (which would then result in `{{ ref('test_' + database) }}`).

### Steps To Reproduce
**Situation 1:**
- Set the target database to `db1`.
- Create two models: `test_db1` and `model_x`.
- Add this reference to `model_x`: `{{ ref('test_' + database) }}`
- Run `dbt compile model_x`

**Situation 2:**
- Set the target database to `db1`.
- Create three models: `test_db1`, `test_db2`, and `model_x`.
- In `dbt_project.yml`, configure `model_x` to use the database `db2`.
- Add this reference to `model_x`: `{{ ref('test_' + database) }}`
- Run `dbt compile model_x`

**Situation 3:**
- Set the target database to `dbz`.
- Create two models: `test_db2` and `model_x`.
- In `dbt_project.yml`, configure `model_x` to use the database `db2`.
- Add this reference to `model_x`: `{{ ref('test_' + database) }}`
- Run `dbt compile model_x`

### Expected behavior
Custom databases & schemas should be resolved before the parsing/manifest phase begins, to ensure that a model's `database` and `schema` remain consistent during dbt's entire execution.

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [ ] bigquery
- [x] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
installed version: 0.18.1
   latest version: 0.18.1

Up to date!

Plugins:
  - snowflake: 0.18.1
```

**The operating system you're using:**
```
macOS Catalina
Version 10.15.7 (19H15)
```

**The output of `python --version`:**
```
Python 3.8.6
```

**Comment by jtcohen6** [comment]
@mroy-seedbox Thanks for the astute outline of the problem, and for the reproduction cases. As you rightly note, dbt does fill in certain "placeholder" values during parsing + manifest construction. In the case of custom schemas and databases, it uses the `target` values; it the case of references, it defaults to the current node reference during parsing, which can result in unexpected behavior (https://github.com/fishtown-analytics/dbt/issues/2793).

> Custom databases & schemas should be resolved before the parsing/manifest phase begins, to ensure that a model's `database` and `schema` remain consistent during dbt's entire execution.

While I agree that this would resolve the current issue, I'm not sure that it's possible: What about models whose `database` and `schema` are dynamically determined, based on a Jinja expression or macro (including `generate_database_name`) that include conditional statements, some of which may depend on the results of an upstream model? Those can only be evaluated during the compilation + run step. When dbt is parsing the model file and tries to generate the DAG, it needs to know about all possible dependencies.

What would be most helpful is to hear from you about the real-world implications of these counterintuitive scenarios. What is it you're trying to do? Are you materializing a series of otherwise-identical models into custom databases, and you want those models to select their parents based on their own configured value of `database`?

Although I agree that situations 2 + 3 are corner cases, I think both have reasonable workarounds using other dbt conventions and constructs:

In situation 2:

```sql
-- may depend on: {{ ref('test_db1') + ref('test_db2') }}
select * from {{ ref('test_' + database) }}
```

In situation 3:

```yml
vars:
  other_db: 'db2'

models:
  my_project:
    model_x: "{{ var('other_db', 'db2') }}"
```
```sql
select * from {{ ref('test_' + var('other_db')) }}
```

**Comment by matt-winkler** [comment]
@jtcohen6 Had an interesting question on this today. . . in dbt Cloud, would making the `var('other_db')` able to respond to successive versions of a database in Snowflake require updating a `vars` block in `dbt_project.yml`? For example, we might have a database named dev_<product>_<version>, where the <version> bit changes over time, either as a timestamp or as a semantic version. This construct could be used to compare different versions of objects within those databases in a dev scenario, row counts and such. Curious to get your thoughts on the above and other patterns that might enable a similar workflow.

**Comment by jtcohen6** [comment]
> in dbt Cloud, would making the `var('other_db')` able to respond to successive versions of a database in Snowflake require updating a `vars` block in `dbt_project.yml`? 

@matt-winkler I don't think so! Passing in that variable from the command arg (`--vars 'other_db: dev_1234'`) should work just fine. In fact, I don't think it's necessary to define the `vars` block in `dbt_project.yml` at all, so long as there's a default value passed each time the variable is called (`{{ var('other_db', 'db2') }}`).

**Comment by mroy-seedbox** [comment]
Well, years later, I figured out that this is actually possible when using more static values which are initialized from the very first parsing of the model, such as `model.alias` and `model.fqn`. 🎉 🍾  🚀 

Whereas `model.database` and `model.schema` are not initialized early enough, and their values can change from the profile definition to their custom values later on.
